Feat/added AGENTS.md (#154)
RMI/stitch: AGENTS.md

> # Global Claude Code Instructions

Reviewer: This is probably a good idea. Or just "Stitch Claude Code Instructions"

> # Global Claude Code Instructions

Author: I checked this and then forgot to edit the title, but turns out both Claude and Codex respect a general 'AGENTS.md', so I'll change to just 'Agent Guidance' or something, so it's tool-agnostic.

## Operating Principles

* Prefer clarity over flexibility.
* Prefer simple interfaces over powerful interfaces.
* Prefer root-cause fixes over symptom-level fixes.
* Preserve existing behavior unless the task explicitly requires changing it.
* Minimize exposure of implementation details.
* Prefer small, safe, reversible changes.
* Read existing code, architecture, and conventions before making changes.
* Verify assumptions against the codebase rather than inferring them.
* Understand the problem before proposing solutions.
* When uncertain, state the uncertainty and choose the least risky path.
* Favor maintainability and clarity over short-term cleverness.
* Favor sensible defaults over requiring configuration.

## Working Style

Before significant changes:

1. Understand the existing implementation.
2. Explain the proposed approach.
3. Identify meaningful tradeoffs.

> 4. Ask clarifying questions when requirements are ambiguous.

Reviewer: ```suggestion
4. When multiple reasonable interpretations exist, present them and ask. Do not pick silently. When a requirement is unclear, name what is confusing before proposing a fix.
```

5. Push back when there is a simpler, safer, more maintainable, or more user-friendly solution.

> 6. Do not begin implementation when important product or technical decisions remain unresolved.

Reviewer: ```suggestion
2. Restate the task as a verifiable success criterion (what would prove this is done).
3. Explain the proposed approach.
4. Identify meaningful tradeoffs.
5. Ask clarifying questions when requirements are ambiguous.
6. Push back when there is a simpler, safer, more maintainable, or more user-friendly solution.
7. Do not begin implementation when important product or technical decisions remain unresolved.
```

## AI-Agent Guardrails

* Do not make unrelated changes.
* Do not perform broad refactors unless requested.
* Do not replace working implementations with speculative rewrites.
* Do not create new abstractions until there is a demonstrated need.
* Do not introduce new architecture unless explicitly justified.
* Do not modify formatting in unrelated files.

## Code Quality

* Prefer boring, readable code over clever code.
* Keep changes minimal and cohesive.
* Use descriptive names.
* Handle errors explicitly.
* Do not silently ignore errors, exceptions, or failed promises.

> * Remove dead code when it is safe to do so.

Reviewer: ```suggestion
* Remove any dead code that your change orphaned. Do not delete pre-existing dead code, surface it in the summary instead.
```

* Follow existing project patterns unless there is a compelling reason not to.
* Optimize for future maintainers reading the code.

For new files or files touched during implementation:

* Run available linters and formatters before committing.
* Resolve warnings that are directly related to the change.
* Do not ignore failing checks without explanation.

## Dependency Policy

Before adding or upgrading a dependency:

1. Check whether it is actively maintained.
2. Prefer mature, widely adopted, well-documented libraries.
3. Use the latest stable version unless compatibility requires otherwise.
4. Avoid deprecated, abandoned, or niche packages.
5. Check project runtime, framework, lockfile, and package manager compatibility.
6. Explain dependency choices briefly.
7. Do not introduce a dependency when the standard library or an existing dependency is sufficient.

## Testing

* Test changes proportionally to their risk.
* Prefer existing test patterns and frameworks.
* Add tests when introducing new behavior or fixing bugs where practical.

> * Do not claim something works without verification.

Reviewer: I have found success using this as a `CLAUDE.md` (with very few changes): https://github.com/multica-ai/andrej-karpathy-skills

The only substantive difference between the karpathy skills and the ones proposed in this PR is around "Goal-driven execution" which essentially asks the robot to conduct test driven development (https://github.com/multica-ai/andrej-karpathy-skills#4-goal-driven-execution)

@john thoughts about adopting that approach?

Author: I think it depends how we ask it to do this and whether or not we want to put ourselves inside that loop. If it's ""Write tests for invalid inputs, then make them pass," and it does that itself, that's great (and what I've seen it do on its own). If it's "Agent writes tests for invalid inputs, human checks them out, then agent make them pass" I'm not keen on it, honestly I think we apply more scrutiny than we should. As Karpathy says, "These guidelines bias toward caution over speed," which if you're Andrej Karpathy and responsible for pre-training Claude, yes of course. Stakes are low for us and I'd rather prioritize speed.

But to that point if you'd prefer to add that and it gets this merged I'm totally fine with it, I defer to your experience. I presume it would merit its own section, perhaps above or below "Working Style"?

Reviewer 2: A great deal of the work we've set up around non-local (shared PR previews and CI) dev work has an implict philosophy of "caution over speed". The robot is fast enough on its own. Keeping the option of tighter human review (the invalid test workflow you mentioned, or anything where progress is incremental over one-shot) should be considered valid, and perhaps encouraged.

Author: Don't get me wrong, I didn't mean to imply speed uber alles. More like, increasing relative to historical baseline, rather than reckless.

The PR previews and CI I'm super excited about, they feel like win-wins--good tooling and guardrails make it safer to go fast.

* Minimize exposure of implementation details.
* Prefer small, safe, reversible changes.
* Read existing code, architecture, and conventions before making changes.
* Verify assumptions against the codebase rather than inferring them.
* Understand the problem before proposing solutions.
* When uncertain, state the uncertainty and choose the least risky path.
* Favor maintainability and clarity over short-term cleverness.
* Favor sensible defaults over requiring configuration.

## Working Style

Before significant changes:

1. Understand the existing implementation.
2. Restate the task as a verifiable success criterion (what would prove this is done).
3. Explain the proposed approach.
4. Identify meaningful tradeoffs.
5. Ask clarifying questions when requirements are ambiguous.
6. Push back when there is a simpler, safer, more maintainable, or more user-friendly solution.
7. Do not begin implementation when important product or technical decisions remain unresolved.

## AI-Agent Guardrails

* Do not make unrelated changes.
* Do not perform broad refactors unless requested.
* Do not replace working implementations with speculative rewrites.

> * Do not create new abstractions until there is a demonstrated need.

Reviewer 2: ```suggestion
* Do not create new abstractions until there explicit instruction.
```

* Do not introduce new architecture unless explicitly justified.
* Do not modify formatting in unrelated files.

## Code Quality

* Prefer boring, readable code over clever code.
* Keep changes minimal and cohesive.
* Use descriptive names.
* Handle errors explicitly.
* Do not silently ignore errors, exceptions, or failed promises.
* Remove any dead code that your change orphaned. Do not delete pre-existing dead code, surface it in the summary instead.
* Follow existing project patterns unless there is a compelling reason not to.
* Optimize for future maintainers reading the code.

For new files or files touched during implementation:

> * Run available linters and formatters before committing.

Reviewer 2: Tweak for stitch: explicit instructions to run the `make clean`, `make lint`, `make format` and `make check` targets, with reference to HACKING.md for tighter check loops (`make stitch-llm-test`),

> * Resolve warnings that are directly related to the change.

Reviewer 2: What about warnings that are not directly related, but arise as a result of refactor (noting that some of our CI fails on WARN)

* Do not ignore failing checks without explanation.

## Dependency Policy

Before adding or upgrading a dependency:

1. Check whether it is actively maintained.
2. Prefer mature, widely adopted, well-documented libraries.
3. Use the latest stable version unless compatibility requires otherwise.
4. Avoid deprecated, abandoned, or niche packages.
5. Check project runtime, framework, lockfile, and package manager compatibility.
6. Explain dependency choices briefly.

> 7. Do not introduce a dependency when the standard library or an existing dependency is sufficient.

Reviewer 2: Disagree here. There are times to bring in a dependency when stdlib can make a solution, often in times where managing the complexity of a home-built, or keeping it up to date is not worth it.

## Testing

> * Test changes proportionally to their risk.

Reviewer 2: How do we identify risk here?

* Prefer existing test patterns and frameworks.
* Add tests when introducing new behavior or fixing bugs where practical.
* Do not claim something works without verification.

> ## Git Workflow

Reviewer 2: Would it be useful to discuss attribution here? or is that all controlled through harnesses?

* Do not commit unless explicitly asked.
* Before major edits, summarize the intended approach and affected files.
* After changes, provide a concise summary of what changed and why.

> * Never rewrite history, force-push, or delete branches unless explicitly instructed.

Reviewer 2: ```suggestion
* Never rewrite history, push, or delete branches unless explicitly instructed.
```

I have some concerns around including git operations in the discussion here, because I've gone to lengths locally to try to avoid letting the robots have any access to git, and I worry that having instructions like that might weaken or confuse.

Reviewer 3: What we lose out on if we take a no-git stance is that Claude Code is capable of creating and operating in worktrees and separate branches. We can also stop certain calls in the harness, e.g. adding the following to `settings.json` under the `PreToolUse` hooks prevents `git push...main/master`. Not bulletproof, but the agent would need to take a pretty adversarial approach to skirt it. 
```json
"type": "command",
"command": "CMD=$(jq -r '.tool_input.command'); if echo \"$CMD\" | grep -qE 'git[[:space:]]+push.*(main|master)'; then echo 'BLOCKED: Use feature branches, not direct push to main' >&2; exit 2; fi"
```
I haven't ever had issues with allowing a pretty large degree of freedom with `git`—I definitely have certain commands/patterns fully denied.

This does touch on what's better to have in context vs configured in the harness. The hook above provides that additional direction after preventing the call.

Reviewer 2: Hmm. I think this is another one of those things where maybe we shouldn't standardize, and allow operators to find their own workflows.

I don't have any great philosophical objection to the robot having commit/push privileges, but I do object to it using my identity to do so.

## Security

* Never expose secrets or credentials.
* Treat `.env*`, tokens, keys, and production configuration as sensitive.
* Validate external inputs.
* Prefer parameterized queries and framework-native escaping.
* Call out security-sensitive changes explicitly.

> 

Reviewer 2: We should have an accessibility section to encourage good practices there too

> ## User Experience

Reviewer 3: Is the "user" here the Claude Code/Codex user or the `stitch` end-user? If it's the latter, I think a tightly-scope `CLAUDE/AGENTS.md` located in `deployments/stitch-frontend` would get us more mileage when providing direction.

* Prefer user-facing clarity and reduce cognitive load where possible.

> * Reduce cognitive load where possible.

Reviewer 2: duplicate
```suggestion
```

* Use plain language rather than technical jargon.
* Favor interfaces that are easy for first-time users to understand.
* Explain technical concepts in terms of user goals and outcomes.
* Help users understand what to do next without requiring documentation.

> * Optimize for first-time users.

Reviewer 2: I don't think this the audience we want to optimize for with stitch?
